src/volatile_set.c

```diff
@@ -1,79 +1,1162 @@
 #include <string.h>
 #include "volatile_set.h"
+#include "rax.h"
 #include "zmalloc.h"
-#include "config.h"
 #include "endianconv.h"
 #include "serverassert.h"
+#include "hashtable.h"
+#include "server.h"
+#include <stdint.h>
 
-#define EXPIRY_HASH_SIZE 16
-volatile_set *createVolatileSet(volatileEntryType *type) {
-    volatile_set *set = zmalloc(sizeof(volatile_set));
-    set->etypr = type;
-    set->expiry_buckets = raxNew();
-    return set;
+/*************************************************************************************************************
+ *                                pointer_vector Implementation
+ *************************************************************************************************************/
+
+#define PV_CARD_BITS 30
+#define PV_ALLOC_BITS 34
+#define PV_MAX_ELEMENTS ((1ULL << PV_CARD_BITS) - 1)
+#define PV_HEADER_SIZE (sizeof(pointer_vector))
+#define PV_ELEM_SIZE (sizeof(void *))
+#define PV_ALLOC(pv) (pv ? pv->alloc : 0)
+#define PV_LEN(pv) (pv ? pv->len : 0)
+#define PV_USED_SIZE(pv) (pv ? (PV_HEADER_SIZE + (pv_len(pv)) * PV_ELEM_SIZE) : 0)
+
+/* Custom vector structure with embedded allocation and length counters */
+typedef struct {
+    uint64_t len : 30;   /* Number of elements */
+    uint64_t alloc : 34; /* Allocated capacity */
+    void *data[];        /* Flexible array member */
+} pointer_vector;
+
+/* Returns the number of elements currently stored in the pointer_vector.
+ *
+ * Arguments:
+ *   vec - The pointer_vector to query.
+ *
+ * Return:
+ *   The number of elements in the vector.
+ *   Note that a NULL is a !!!valid!!! vector - returns 0 if the vector is NULL. */
+static inline uint32_t pv_len(pointer_vector *vec) {
+    return PV_LEN(vec);
+}
+
+/* Ensures that a pointer_vector has enough capacity to hold additional elements.
+ *
+ * This function guarantees that the given pointer_vector `sv` has at least enough
+ * allocated space to accommodate `capacity` more elements, growing it if necessary.
+ * If the vector is currently `NULL`, it will be newly allocated.
+ *
+ * The allocation is handled using `zmalloc` or `zrealloc_usable`, depending on whether
+ * the vector is new or already initialized. The internal `alloc` field is updated to
+ * reflect the actual allocated size.
+ *
+ * Arguments:
+ *   sv       - Pointer to an existing pointer_vector or NULL.
+ *   capacity - The number of additional elements the vector should be able to accommodate.
+ *
+ * Return:
+ *   A pointer to the resized (or newly allocated) pointer_vector with sufficient capacity.
+ *   Returns NULL only if the allocation fails.
+ *
+ * Note:
+ *   The `capacity` is the number of *additional* elements beyond the current length.
+ *   This function does not modify the vector's logical length (`len`), only its allocation. */
+pointer_vector *pv_grow_to_fit(pointer_vector *sv, size_t capacity) {
```

Comment:
* It's confusing to use the word "capacity" for the additional number of elements to reserve. Capacity typically means the total capacity, unless we say a word like additional or delta.

* Use camel case for function name. Using "pv" as the prefix for these functions seems good to me.

* It's internal to the file, right? Then also make it `static`.

* The variable name `sv` for pointer vector seems odd. Use `pv` or `vec` instead.

* To match the sds internal naming, have you considered the name `pvMakeRoomFor`?

Reply:
noted. I will fix